File: DaS.ScriptLib.CS/Game/Data/Structures/Loc.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using System.Threading.Tasks;
namespace DaS.ScriptLib.Game.Data.Structures
{
	public class Loc
	{
		public Vec3 Pos;

		public Heading Rot;
		public Loc()
		{
			Pos = Vec3.Zero;
			Rot = new Heading(0);
		}

		public Loc(Vec3 pos, float rot)
		{
			this.Pos = pos;
			this.Rot = new Heading(rot);
		}

		public Loc(Vec3 pos, Heading rot)
		{
			this.Pos = pos;
			this.Rot = rot;
		}

		public Loc(Vec3 pos) : this(pos, 0)
		{
		}

		public Loc(float posX, float posY, float posZ, float rotHeading)
		{
			Pos = new Vec3(posX, posY, posZ);
			this.Rot = new Heading(rotHeading);
		}

		public Loc(posX, posY, posZ)
		{
			Pos = new Vec3(posX, posY, posZ);
			Rot = new Heading(0);
		}

		public bool IsZero {
			get { return Pos.X == 0 && Pos.Y == 0 && Pos.Z == 0 && Rot.HeadingValue == 0; }
		}

		public object AngleTo(Loc other)
		{
			return Pos.GetLateralAngleTo(other.Pos);
		}


		public static readonly Loc Zero = new Loc(0, 0, 0, 0);
		public static bool operator !=(Loc left, Loc right)
		{
			return (left().Pos != right().Pos) | (left().Rot.PlanarValue != right().Rot.PlanarValue);
		}

		public static bool operator ==(Loc left, Loc right)
		{
			return (left().Pos == right().Pos) && (left().Rot.PlanarValue == right().Rot.PlanarValue);
		}

	}
}

File: DaS.ScriptLib.CS/Game/Data/Structures/Loc.vb
Namespace Game.Data.Structures
    Public Class Loc
        Public Pos As Vec3
        Public Rot As Heading

        Public Sub New()
            Pos = Vec3.Zero
            Rot = New Heading(0)
        End Sub

        Public Sub New(pos As Vec3, rot As Single)
            Me.Pos = pos
            Me.Rot = New Heading(rot)
        End Sub

        Public Sub New(pos As Vec3, rot As Heading)
            Me.Pos = pos
            Me.Rot = rot
        End Sub

        Public Sub New(pos As Vec3)
            Me.New(pos, 0)
        End Sub

        Public Sub New(posX As Single, posY As Single, posZ As Single, rotHeading As Single)
            Pos = New Vec3(posX, posY, posZ)
            Me.Rot = New Heading(rotHeading)
        End Sub

        Public Sub New(posX, posY, posZ)
            Pos = New Vec3(posX, posY, posZ)
            Rot = New Heading(0)
        End Sub

        Public ReadOnly Property IsZero As Boolean
            Get
                Return Pos.X = 0 AndAlso Pos.Y = 0 AndAlso Pos.Z = 0 AndAlso Rot.HeadingValue = 0
            End Get
        End Property

        Public Function AngleTo(other As Loc)
            Return Pos.GetLateralAngleTo(other.Pos)
        End Function

        Public Shared ReadOnly Zero As New Loc(0, 0, 0, 0)

        Public Shared Operator <>(ByVal left As Loc, ByVal right As Loc) As Boolean
            Return (left.Pos <> right.Pos) Or (left.Rot.PlanarValue <> right.Rot.PlanarValue)
        End Operator

        Public Shared Operator =(ByVal left As Loc, ByVal right As Loc) As Boolean
            Return (left.Pos = right.Pos) AndAlso (left.Rot.PlanarValue = right.Rot.PlanarValue)
        End Operator

    End Class
End Namespace